examples/tonic-key-value-store/src/main.rs

```diff
@@ -0,0 +1,265 @@
+
+            // Read the value from stdin
+            let mut stdin = tokio::io::stdin();
+            let mut value = Vec::new();
+            stdin.read_to_end(&mut value).await.unwrap();
+
+            // Issue a `SetRequest`
+            client.set(SetRequest { key, value }).await.unwrap();
+
+            // All good :+1:
+            println!("OK");
+        }
+    }
+}
+
+// We make this a separate function so we're able to call it from tests.
+async fn serve_forever(listener: TcpListener) -> Result<(), hyper::Error> {
+    // Build our database for holding the key/value pairs
+    let db = Arc::new(RwLock::new(HashMap::new()));
+
+    // Build our tonic `Service`
+    let service = key_value_store_server::KeyValueStoreServer::new(ServerImpl { db });
+
+    // Apply middlewares to our service
+    let service = ServiceBuilder::new()
+        // Set a timeout
+        .timeout(Duration::from_secs(10))
+        // Compress responses
+        .layer(CompressionLayer::new())
+        // Mark the `Authorization` header as sensitive so it doesn't show in logs
+        .layer(SetSensitiveHeaderLayer::new(header::AUTHORIZATION))
+        // Build our final `Service`
+        .service(service);
+
+    // Run the service using hyper
+    let addr = listener.local_addr().unwrap();
+
+    tracing::info!("Listening on {}", addr);
+
+    // We cannot use `tonic::transport::Server` directly as it requires services to implement
+    // `tonic::transport::NamedService` which tower-http middlewares don't
+    Server::from_tcp(listener)
+        .unwrap()
+        .http2_only(true)
+        .serve(Shared::new(service))
+        .await?;
+
+    Ok(())
+}
+
+// Implementation of the server trait generated by tonic
+#[derive(Debug, Clone)]
+struct ServerImpl {
+    db: Arc<RwLock<HashMap<String, Bytes>>>,
+}
+
+#[async_trait]
+impl key_value_store_server::KeyValueStore for ServerImpl {
+    async fn get(&self, request: Request<GetRequest>) -> Result<Response<GetReply>, Status> {
+        let key = request.into_inner().key;
+
+        if let Some(value) = self.db.read().unwrap().get(&key).cloned() {
+            let reply = GetReply {
+                value: value.to_vec(),
+            };
+
+            Ok(Response::new(reply))
+        } else {
+            Err(Status::not_found("key not found"))
+        }
+    }
+
+    async fn set(&self, request: Request<SetRequest>) -> Result<Response<SetReply>, Status> {
+        let SetRequest { key, value } = request.into_inner();
+        let value = Bytes::from(value);
+
+        self.db.write().unwrap().insert(key, value);
+
+        Ok(Response::new(SetReply {}))
+    }
+}
+
+// Our final client type with middlewares applied
+type Client = key_value_store_client::KeyValueStoreClient<
+    Decompression<SetRequestHeader<Channel, HeaderValue>>,
+>;
+
+// Build a client with a few middlewares applied and connect to the server
+async fn make_client(addr: SocketAddr) -> Result<Client, tonic::transport::Error> {
```

Comment:
Something like this

```rust
async fn make_client(
    addr: SocketAddr,
) -> Result<
    KeyValueStoreClient<
        impl Service<
                hyper::Request<BoxBody>,
                Response = hyper::Response<impl HttpBody<Error = impl Into<BoxError>>>,
                Error = impl Into<BoxError>,
            > + Clone
            + Send
            + Sync
            + 'static,
    >,
    tonic::transport::Error,
> {
```

The `Clone + Send + Sync + 'static` bound isn't strictly necessary for this example but I figure it might be nice to include for users copying this code.